From 06a5ba5e6335ee04ec09f4b066925847e870dea7 Mon Sep 17 00:00:00 2001
From: Aishwarya R <aisr@qti.qualcomm.com>
Date: Tue, 22 Apr 2025 10:23:55 +0530
Subject: [PATCH 2/4] UPSTREAM:DP:wifi: mac80211: Enhance RX stats reporting
 per TID specific with drop reasons

Improve the RX statistics reporting in the mac80211 subsystem by adding detailed
counters for various RX packet reasons and drop statistics for specific TIDs.

Below is the debugfs file format will be available for TID 0 - 15 with
all packet reasons and drop reasons

$cat sys/kernel/debug/ieee80211/phy00/netdev:wlan1/txrx_stats_reason

 		MAC80211 RX STATS
 TID 	 packets bytes netif_pkts forwarded_pkts fast_path_pkts multicast_pkts local_stack_pkts queue_pkts monitor_pkts
 0        923      96115    0            0              0              0              923              14025      0
 1        0        0        0            0              0              0              0                0          0
 2        0        0        0            0              0              0              0                0          0
 3        0        0        0            0              0              0              0                0          0
 4        0        0        0            0              0              0              0                0          0
 5        0        0        0            0              0              0              0                0          0
 6        66       2772     0            0              0              0              66               0          0
 7        33       4338     0            0              0              0              33               0          0
 ....
 15       0        0        0            0              0              0              0                0          0

 		MAC80211 RX DROP STATS
TID 0	38: 13	,
TID 1
TID 2
TID 3

....
TID15

Patch-dependency: none
Signed-off-by: Aishwarya R <aisr@qti.qualcomm.com>
---
 include/net/mac80211.h        |   2 +
 net/mac80211/debugfs_netdev.c | 237 +++++++++++++++++++++++++++++++++-
 net/mac80211/drop.h           |  60 +++++++++
 net/mac80211/ieee80211_i.h    |  52 ++++++++
 net/mac80211/rx.c             | 212 ++++++++++++++++++++++++++----
 net/mac80211/wpa.c            |   1 -
 6 files changed, 535 insertions(+), 29 deletions(-)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 8187e70..f0c6aa1 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1266,6 +1266,7 @@ struct ieee80211_tx_info {
 			u8 pad[36];
 			u32 nw_delay;
 		} sawf;
+		u8 tid;
 	};
 };
 
@@ -1670,6 +1671,7 @@ struct ieee80211_rx_status {
 	u8 zero_length_psdu_type;
 	u8 link_valid:1, link_id:4;
 	u8 eht_num_user;
+	u8 tid;
 };
 
 static inline u32
diff --git a/net/mac80211/debugfs_netdev.c b/net/mac80211/debugfs_netdev.c
index 149897e..4fbb675 100644
--- a/net/mac80211/debugfs_netdev.c
+++ b/net/mac80211/debugfs_netdev.c
@@ -22,6 +22,71 @@
 #include "debugfs_netdev.h"
 #include "driver-ops.h"
 
+const char *rx_drop_reason_strings[RX_DROP_REASON_MAX + TX_DROP_REASON_MAX] = {
+	[RX_DROP_MIC_FAIL] = "mic_fail",
+	[RX_DROP_REPLAY] = "replay",
+	[RX_DROP_BAD_MMIE] = "bad_mmie",
+	[RX_DROP_DUP] = "dup",
+	[RX_DROP_SPURIOUS] = "spurious",
+	[RX_DROP_DECRYPT_FAIL] = "decrypt_fail",
+	[RX_DROP_NO_KEY_ID] = "no_keyid",
+	[RX_DROP_BAD_CIPHER] = "bad_cipher",
+	[RX_DROP_OOM] = "OOM",
+	[RX_DROP_NONSEQ_PN] = "nonseq_pn",
+	[RX_DROP_BAD_KEY_COLOR] = "bad_key_color",
+	[RX_DROP_BAD_4ADDR] = "bad_4addr",
+	[RX_DROP_BAD_AMSDU] = "bad_amsdu",
+	[RX_DROP_BAD_AMSDU_CIPHER] = "Bad amsdu cipher",
+	[RX_DROP_INVALID_8023] = "Invalid 8023",
+	[RX_DROP_RUNT_ACTION] = "Runt action",
+	[RX_DROP_UNPROT_ACTION] = "Unprot action",
+	[RX_DROP_ACTION_UNKNOWN_SRC] = "Unknown src",
+	[RX_DROP_REJECTED_ACTION_RESPONSE] = "Rej action resp",
+	[RX_DROP_EXPECT_DEFRAG_PROT] = "defrag port",
+	[RX_DROP_WEP_DEC_FAIL] = "wep dec fail",
+	[RX_DROP_NO_IV] = "No IV",
+	[RX_DROP_NO_ICV] = "No ICV",
+	[RX_DROP_AP_RX_GROUPCAST] = "RX groupcast",
+	[RX_DROP_SHORT_MMIC] = "short mmic",
+	[RX_DROP_MMIC_FAIL] = "mmic fail",
+        [RX_DROP_SHORT_TKIP] = "short tkip",
+	[RX_DROP_TKIP_FAIL] = "tkip fail",
+	[RX_DROP_SHORT_CCMP] = "short ccmp",
+	[RX_DROP_SHORT_CCMP_MIC] = "short ccmp mic",
+	[RX_DROP_SHORT_GCMP] = "short gcmp",
+	[RX_DROP_SHORT_GCMP_MIC] = "short gcmp mic",
+	[RX_DROP_SHORT_CMAC] = "short cmac",
+	[RX_DROP_SHORT_CMAC256] = "cmac256",
+	[RX_DROP_SHORT_GMAC] = "short gmac",
+	[RX_DROP_UNEXPECTED_4ADDR_FRAME] = "unexpected 4addr",
+	[RX_DROP_BAD_BCN_KEYIDX] = "bcn keyidx",
+	[RX_DROP_BAD_MGMT_KEYIDX] = "mgmt keyidx",
+	[RX_DROP_INVALID_SKB] = "invalid skb",
+	[RX_DROP_SHORT_HDR] = "short hdr",
+	[RX_DROP_NO_PUBSTA] = "no pubsta",
+	[RX_DROP_NO_SET_STA] = "no set sta",
+	[RX_DROP_NO_FAST_RX] = "no fast rx",
+	[RX_DROP_NO_LINK_STA] = "no link sta",
+	[RX_DROP_FRAME_ERR] = "frame_err",
+};
+
+const char *tx_drop_reason_strings[TX_DROP_REASON_MAX] = {
+	[TX_DROP_SDATA_STATE] = "sdata_state",
+	[TX_DROP_SKB_SANITY_CHECK_FAIL] = "skb_sanity_check_fail",
+	[TX_DROP_SKB_RESIZE_FAIL] = "skb_resize_fail",
+	[TX_DROP_SKB_QUEUE_FULL] = "skb_queue full",
+	[TX_DROP_RA_STA_FAIL] = "ra_sta_failure",
+	[TX_DROP_HW_CHECK_FAIL] = "hw_check_failure",
+	[TX_DROP_TX_SKB_FIXUP_FAIL] = "skb_fixup_failure",
+	[TX_DROP_BAD_LINK_ID] = "bad_link_id",
+	[TX_DROP_BAD_CHANCTX_CONF] = "bad_chanctx_conf",
+	[TX_DROP_BAD_LINK_CONF] = "bad_link_conf",
+	[TX_DROP_UNKNOWN_VIF_TYPE] = "unknown_vif_type",
+	[TX_DROP_UNAUTHORIZED_STA] = "unauthorized_sta",
+	[TX_DROP_QUEUE_PURGE] = "queue_purge",
+	[TX_DROP_MISC] = "misc"
+};
+
 static ssize_t ieee80211_if_read(
 	void *data,
 	char __user *userbuf,
@@ -255,6 +320,175 @@ ieee80211_if_fmt_hw_queues(const struct ieee80211_sub_if_data *sdata,
 }
 IEEE80211_IF_FILE_R(hw_queues);
 
+static ssize_t ieee80211_if_fmt_txrx_stats_reason_read(struct file *file, char __user *userbuf,
+						       size_t count, loff_t *ppos)
+{
+	struct ieee80211_sub_if_data *sdata = file->private_data;
+	struct txrx_stats *total_stats;
+	u8 tid, reason;
+	int cpu;
+	const int size = 15000;
+	int len = 0, retval;
+	char *buf;
+
+	buf = vmalloc(size);
+	if (!buf)
+		return -ENOMEM;
+
+	total_stats = kzalloc(sizeof(*total_stats), GFP_KERNEL);
+	if (!total_stats) {
+		vfree(buf);
+		return -ENOMEM;
+	}
+
+	len = scnprintf(buf + len, size - len, "\n\t\tMAC80211 RX STATS\t\t\n");
+
+	len += scnprintf(buf + len, size - len, "TID \t packets bytes netif_pkts forwarded_pkts fast_path_pkts multicast_pkts local_stack_pkts queue_pkts monitor_pkts\n");
+
+	for_each_possible_cpu(cpu) {
+		struct pcpu_txrx_stats *txrx_stats = per_cpu_ptr(sdata->txrx_stats, cpu);
+		u64_stats_update_begin(&txrx_stats->syncp);
+
+		for (u8 tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
+			total_stats->tid_stats[tid].rx_packets += txrx_stats->tid_stats[tid].rx_packets;
+			total_stats->tid_stats[tid].rx_bytes += txrx_stats->tid_stats[tid].rx_bytes;
+			total_stats->tid_stats[tid].rx_netif_pkts += txrx_stats->tid_stats[tid].rx_netif_pkts;
+			total_stats->tid_stats[tid].rx_forwarded_pkts += txrx_stats->tid_stats[tid].rx_forwarded_pkts;
+			total_stats->tid_stats[tid].rx_fast_path_pkts += txrx_stats->tid_stats[tid].rx_fast_path_pkts;
+			total_stats->tid_stats[tid].rx_multicast_pkts += txrx_stats->tid_stats[tid].rx_multicast_pkts;
+			total_stats->tid_stats[tid].rx_local_stack_pkts += txrx_stats->tid_stats[tid].rx_local_stack_pkts;
+			total_stats->tid_stats[tid].rx_queue_pkts += txrx_stats->tid_stats[tid].rx_queue_pkts;
+			total_stats->tid_stats[tid].rx_monitor_pkts += txrx_stats->tid_stats[tid].rx_monitor_pkts;
+			total_stats->tid_stats[tid].tx_eth_pkts += txrx_stats->tid_stats[tid].tx_eth_pkts;
+			total_stats->tid_stats[tid].tx_eth_pkts_bytes += txrx_stats->tid_stats[tid].tx_eth_pkts_bytes;
+			total_stats->tid_stats[tid].fast_tx_pkts += txrx_stats->tid_stats[tid].fast_tx_pkts;
+			total_stats->tid_stats[tid].fast_tx_pkts_bytes += txrx_stats->tid_stats[tid].fast_tx_pkts_bytes;
+			total_stats->tid_stats[tid].tx_nwifi_pkts += txrx_stats->tid_stats[tid].tx_nwifi_pkts;
+			total_stats->tid_stats[tid].tx_nwifi_pkts_bytes += txrx_stats->tid_stats[tid].tx_nwifi_pkts_bytes;
+			total_stats->tid_stats[tid].tx_monitor_pkts += txrx_stats->tid_stats[tid].tx_monitor_pkts;
+			total_stats->tid_stats[tid].tx_monitor_pkts_bytes += txrx_stats->tid_stats[tid].tx_monitor_pkts_bytes;
+			total_stats->tid_stats[tid].tx_multicast_pkts += txrx_stats->tid_stats[tid].tx_multicast_pkts;
+			total_stats->tid_stats[tid].tx_multicast_pkts_bytes += txrx_stats->tid_stats[tid].tx_multicast_pkts_bytes;
+
+			for (u8 reason = 0; reason < RX_DROP_REASON_MAX; reason++) {
+				total_stats->tid_stats[tid].rx_drop_stats[reason] += txrx_stats->tid_stats[tid].rx_drop_stats[reason];
+			}
+
+			for (u8 reason = 0; reason < TX_DROP_REASON_MAX; reason++) {
+				total_stats->tid_stats[tid].tx_drop_stats[reason] += txrx_stats->tid_stats[tid].tx_drop_stats[reason];
+			}
+		}
+		u64_stats_update_end(&txrx_stats->syncp);
+	}
+
+	for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
+		len += scnprintf(buf + len, size - len, "%-8d %-8llu %-8llu %-12llu %-14llu %-14llu %-14llu %-16llu %-10llu %-12llu\n", tid,
+				 total_stats->tid_stats[tid].rx_packets,
+				 total_stats->tid_stats[tid].rx_bytes,
+				 total_stats->tid_stats[tid].rx_netif_pkts,
+				 total_stats->tid_stats[tid].rx_forwarded_pkts,
+				 total_stats->tid_stats[tid].rx_fast_path_pkts,
+				 total_stats->tid_stats[tid].rx_multicast_pkts,
+				 total_stats->tid_stats[tid].rx_local_stack_pkts,
+				 total_stats->tid_stats[tid].rx_queue_pkts,
+				 total_stats->tid_stats[tid].rx_monitor_pkts);
+	}
+
+	len += scnprintf(buf + len, size - len, "\n\t\tMAC80211 RX DROP STATS\t\t\n");
+
+	for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
+		len += scnprintf(buf + len, size - len, "TID %d\t", tid);
+		for (reason = 0; reason < RX_DROP_REASON_MAX; reason++) {
+			if (total_stats->tid_stats[tid].rx_drop_stats[reason])
+				len += scnprintf(buf + len, size - len,"%d: %llu\t,",
+						 reason, total_stats->tid_stats[tid].rx_drop_stats[reason]);
+		}
+		len += scnprintf(buf + len, size - len, "\n");
+	}
+
+	len += scnprintf(buf + len, size - len, " \n\t\tMAC80211 TX STATS\t\t\n");
+
+#ifdef CPTCFG_MAC80211_DEBUG_COUNTERS
+	len += scnprintf(buf + len, size - len,
+			 "TransmittedFragmentCount: %llu \n"
+			 "MulticastTransmittedFrameCount: %llu \n"
+			 "FailedCount: %llu \n"
+			 "RetryCount: %llu \n"
+			 "MultipleRetryCount: %llu \n"
+			 "FrameDuplicateCount: %llu \n"
+			 "TransmittedFrameCount: %llu \n",
+			 sdata->local->dot11TransmittedFragmentCount,
+			 sdata->local->dot11MulticastTransmittedFrameCount,
+			 sdata->local->dot11FailedCount,
+			 sdata->local->dot11RetryCount,
+			 sdata->local->dot11MultipleRetryCount,
+			 sdata->local->dot11FrameDuplicateCount,
+			 sdata->local->dot11TransmittedFrameCount);
+#endif
+	len += scnprintf(buf + len, size - len, "TID\t packets  bytes  eth_pkts  eth_bytes  fast_tx_pkts  fast_tx_bytes  nwifi_pkts  nwifi_bytes mon_pkts  mon_bytes  mc_pkts  mc_bytes\n");
+
+	for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
+		len += scnprintf(buf + len, size - len, "%-8d %-8llu %-8llu %-10llu %-10llu %-12llu %-12llu %-10llu %-10llu %-10llu %-10llu %-10llu %-10llu\n",
+				 tid,
+				 (total_stats->tid_stats[tid].tx_eth_pkts + total_stats->tid_stats[tid].fast_tx_pkts + total_stats->tid_stats[tid].tx_nwifi_pkts + total_stats->tid_stats[tid].tx_monitor_pkts),
+				 (total_stats->tid_stats[tid].tx_eth_pkts_bytes + total_stats->tid_stats[tid].fast_tx_pkts_bytes + total_stats->tid_stats[tid].tx_nwifi_pkts_bytes + total_stats->tid_stats[tid].tx_monitor_pkts_bytes),
+				 total_stats->tid_stats[tid].tx_eth_pkts,
+				 total_stats->tid_stats[tid].tx_eth_pkts_bytes,
+				 total_stats->tid_stats[tid].fast_tx_pkts,
+				 total_stats->tid_stats[tid].fast_tx_pkts_bytes,
+				 total_stats->tid_stats[tid].tx_nwifi_pkts,
+				 total_stats->tid_stats[tid].tx_nwifi_pkts_bytes,
+				 total_stats->tid_stats[tid].tx_monitor_pkts,
+				 total_stats->tid_stats[tid].tx_monitor_pkts_bytes,
+				 total_stats->tid_stats[tid].tx_multicast_pkts,
+				 total_stats->tid_stats[tid].tx_multicast_pkts_bytes);
+	}
+
+	len += scnprintf(buf + len, size - len, " \n\t\tMAC80211 TX TID DROP STATS\t\t\n");
+
+	for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
+		len += scnprintf(buf + len, size - len, "TID %d\t", tid);
+		for (reason = 0; reason < TX_DROP_REASON_MAX; reason++) {
+		       if (total_stats->tid_stats[tid].tx_drop_stats[reason])
+		       		len += scnprintf(buf + len, size - len, "%d: %llu\t,",
+						 reason, total_stats->tid_stats[tid].tx_drop_stats[reason]);
+		}
+		len += scnprintf(buf + len, size - len, "\n");
+	}
+
+	if (len > size)
+		len = size;
+
+	retval = simple_read_from_buffer(userbuf, count, ppos, buf, len);
+	vfree(buf);
+	kfree(total_stats);
+
+	return retval;
+}
+
+static const struct file_operations txrx_stats_reason_ops = {
+	.read = ieee80211_if_fmt_txrx_stats_reason_read,
+	.open = simple_open,
+	.llseek = generic_file_llseek,
+};
+
+static ssize_t ieee80211_if_parse_reset_txrx_stats(struct ieee80211_sub_if_data *sdata, const char *buf, int buflen)
+{
+	int cpu;
+	u8 tid;
+
+	for_each_possible_cpu(cpu) {
+		struct pcpu_txrx_stats *txrx_stats = per_cpu_ptr(sdata->txrx_stats, cpu);
+
+		u64_stats_update_begin(&txrx_stats->syncp);
+		for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++)
+			memset(&txrx_stats->tid_stats[tid], 0, sizeof(struct txrx_tid_stats));
+		u64_stats_update_end(&txrx_stats->syncp);
+	}
+	return buflen;
+}
+IEEE80211_IF_FILE_W(reset_txrx_stats);
+
 /* STA attributes */
 IEEE80211_IF_FILE(bssid, deflink.u.mgd.bssid, MAC);
 IEEE80211_IF_FILE(aid, vif.cfg.aid, DEC);
@@ -769,7 +1003,8 @@ static void add_common_files(struct ieee80211_sub_if_data *sdata)
 	DEBUGFS_ADD(rc_rateidx_vht_mcs_mask_2ghz);
 	DEBUGFS_ADD(rc_rateidx_vht_mcs_mask_5ghz);
 	DEBUGFS_ADD(hw_queues);
-
+	DEBUGFS_ADD(txrx_stats_reason);
+	DEBUGFS_ADD(reset_txrx_stats);
 	if (sdata->vif.type != NL80211_IFTYPE_P2P_DEVICE &&
 	    sdata->vif.type != NL80211_IFTYPE_NAN)
 		DEBUGFS_ADD(aqm);
diff --git a/net/mac80211/drop.h b/net/mac80211/drop.h
index 725a07a..08d8c57 100644
--- a/net/mac80211/drop.h
+++ b/net/mac80211/drop.h
@@ -84,6 +84,66 @@ enum mac80211_drop_reason {
 #undef DEF
 };
 
+enum rx_pkt_type {
+	RX_NETIF_PKTS,
+	RX_FORWARDED_PKTS,
+	RX_FAST_PATH_PKTS,
+	RX_MULTICAST_PKTS,
+	RX_LOCAL_STACK_PKTS,
+	RX_QUEUED_PKTS,
+	RX_MONITOR_PKTS,
+	RX_TOTAL_PKTS
+};
+
+enum rx_drop_reasons {
+	RX_DROP_MIC_FAIL,
+	RX_DROP_REPLAY,
+	RX_DROP_BAD_MMIE,
+	RX_DROP_DUP,
+	RX_DROP_SPURIOUS,
+	RX_DROP_DECRYPT_FAIL,
+	RX_DROP_NO_KEY_ID,
+	RX_DROP_BAD_CIPHER,
+	RX_DROP_OOM,
+	RX_DROP_NONSEQ_PN,
+	RX_DROP_BAD_KEY_COLOR,
+	RX_DROP_BAD_4ADDR,
+	RX_DROP_BAD_AMSDU,
+	RX_DROP_BAD_AMSDU_CIPHER,
+	RX_DROP_INVALID_8023,
+	RX_DROP_RUNT_ACTION,
+	RX_DROP_UNPROT_ACTION,
+	RX_DROP_ACTION_UNKNOWN_SRC,
+	RX_DROP_REJECTED_ACTION_RESPONSE,
+	RX_DROP_EXPECT_DEFRAG_PROT,
+	RX_DROP_WEP_DEC_FAIL,
+	RX_DROP_NO_IV,
+	RX_DROP_NO_ICV,
+	RX_DROP_AP_RX_GROUPCAST,
+	RX_DROP_SHORT_MMIC,
+	RX_DROP_MMIC_FAIL,
+	RX_DROP_SHORT_TKIP,
+	RX_DROP_TKIP_FAIL,
+	RX_DROP_SHORT_CCMP,
+	RX_DROP_SHORT_CCMP_MIC,
+	RX_DROP_SHORT_GCMP,
+	RX_DROP_SHORT_GCMP_MIC,
+	RX_DROP_SHORT_CMAC,
+	RX_DROP_SHORT_CMAC256,
+	RX_DROP_SHORT_GMAC,
+	RX_DROP_UNEXPECTED_4ADDR_FRAME,
+	RX_DROP_BAD_BCN_KEYIDX,
+	RX_DROP_BAD_MGMT_KEYIDX,
+	RX_DROP_INVALID_SKB,
+	RX_DROP_SHORT_HDR,
+	RX_DROP_NO_PUBSTA,
+	RX_DROP_NO_SET_STA,
+	RX_DROP_NO_FAST_RX,
+	RX_DROP_NO_LINK_STA,
+	RX_DROP_FRAME_ERR,
+	RX_DROP_REASON_MAX
+};
+
 #define RX_RES_IS_UNUSABLE(result)	\
 	(((__force u32)(result) & SKB_DROP_REASON_SUBSYS_MASK) == ___RX_DROP_UNUSABLE)
 
diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index 0855d73..0cbdb39 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1043,6 +1043,45 @@ struct ieee80211_link_data {
 #endif
 };
 
+struct txrx_tid_stats {
+	u64	rx_packets;
+	u64	rx_bytes;
+	u64	rx_drop_packets;
+	u64	rx_drop_bytes;
+	u64	rx_netif_pkts;
+	u64	rx_forwarded_pkts;
+	u64	rx_fast_path_pkts;
+	u64	rx_multicast_pkts;
+	u64	rx_local_stack_pkts;
+	u64	rx_queue_pkts;
+	u64	rx_monitor_pkts;
+	u64	rx_drop_stats[RX_DROP_REASON_MAX];
+	u64	tx_packets;
+	u64	tx_bytes;
+	u64	tx_drop_packets;
+	u64	tx_drop_bytes;
+	u64	tx_eth_pkts;
+	u64	tx_eth_pkts_bytes;
+	u64	fast_tx_pkts;
+	u64	fast_tx_pkts_bytes;
+	u64	tx_multicast_pkts;
+	u64	tx_multicast_pkts_bytes;
+	u64	tx_nwifi_pkts;
+	u64	tx_nwifi_pkts_bytes;
+	u64	tx_monitor_pkts;
+	u64	tx_monitor_pkts_bytes;
+	u64	tx_drop_stats[TX_DROP_REASON_MAX];
+};
+
+struct pcpu_txrx_stats {
+	struct txrx_tid_stats	tid_stats[IEEE80211_NUM_TIDS];
+	struct u64_stats_sync   syncp;
+};
+
+struct txrx_stats {
+	struct txrx_tid_stats   tid_stats[IEEE80211_NUM_TIDS];
+};
+
 struct ieee80211_sub_if_data {
 	struct list_head list;
 
@@ -1143,8 +1182,11 @@ struct ieee80211_sub_if_data {
 #ifdef CPTCFG_MAC80211_NSS_SUPPORT
 	struct nss_virt_if_handle *nssctx;
 #endif
+
+	struct pcpu_txrx_stats __percpu	*txrx_stats;
 	u32 tx_dropped;
 	u32 rx_dropped;
+
 	/* must be last, dynamically sized area in this! */
 	struct ieee80211_vif vif;
 };
@@ -2455,6 +2497,16 @@ static inline int ieee80211_ac_from_tid(int tid)
 	return ieee802_1d_to_ac[tid & 7];
 }
 
+static inline void ieee80211_tx_drop_stats(struct ieee80211_sub_if_data *sdata,
+                                           u8 tid, u32 reason)
+{
+	struct pcpu_txrx_stats *txrx_stats = this_cpu_ptr(sdata->txrx_stats);
+
+	u64_stats_update_begin(&txrx_stats->syncp);
+	txrx_stats->tid_stats[tid].tx_drop_stats[reason]++;
+	u64_stats_update_end(&txrx_stats->syncp);
+}
+
 void ieee80211_dynamic_ps_enable_work(struct work_struct *work);
 void ieee80211_dynamic_ps_disable_work(struct work_struct *work);
 void ieee80211_dynamic_ps_timer(struct timer_list *t);
diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 10f6fdb..e215449 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -55,6 +55,50 @@ static inline void ieee80211_rx_stats(struct net_device *dev, u32 len)
 	u64_stats_update_end(&tstats->syncp);
 }
 
+static inline void ieee80211_rx_stats_reason(struct ieee80211_sub_if_data *sdata, u32 len, u8 tid, u32 reason)
+{
+	struct pcpu_txrx_stats *txrx_stats = this_cpu_ptr(sdata->txrx_stats);
+
+	u64_stats_update_begin(&txrx_stats->syncp);
+	switch (reason) {
+	case RX_NETIF_PKTS:
+		txrx_stats->tid_stats[tid].rx_netif_pkts++;
+		break;
+	case RX_FORWARDED_PKTS:
+		txrx_stats->tid_stats[tid].rx_forwarded_pkts++;
+		break;
+	case RX_FAST_PATH_PKTS:
+		txrx_stats->tid_stats[tid].rx_fast_path_pkts++;
+		break;
+	case RX_MULTICAST_PKTS:
+		txrx_stats->tid_stats[tid].rx_multicast_pkts++;
+		break;
+	case RX_LOCAL_STACK_PKTS:
+		txrx_stats->tid_stats[tid].rx_local_stack_pkts++;
+		break;
+	case RX_QUEUED_PKTS:
+		txrx_stats->tid_stats[tid].rx_queue_pkts++;
+		break;
+	case RX_MONITOR_PKTS:
+		txrx_stats->tid_stats[tid].rx_monitor_pkts++;
+		break;
+	case RX_TOTAL_PKTS:
+		txrx_stats->tid_stats[tid].rx_packets++;
+		txrx_stats->tid_stats[tid].rx_bytes += len;
+		break;
+	}
+	u64_stats_update_end(&txrx_stats->syncp);
+}
+
+static inline void ieee80211_rx_drop_stats_reason(struct ieee80211_sub_if_data *sdata, u32 len, u8 tid, u32 reason)
+{
+	 struct pcpu_txrx_stats *txrx_stats = this_cpu_ptr(sdata->txrx_stats);
+
+	u64_stats_update_begin(&txrx_stats->syncp);
+	txrx_stats->tid_stats[tid].rx_drop_stats[reason]++;
+	u64_stats_update_end(&txrx_stats->syncp);
+}
+
 /*
  * monitor mode reception
  *
@@ -875,6 +919,7 @@ ieee80211_rx_monitor(struct ieee80211_local *local, struct sk_buff *origskb,
 	struct ieee80211_sub_if_data *monitor_sdata;
 	bool only_monitor = false;
 	unsigned int min_head_len;
+	bool tid_stats_disable = local->hw.tid_stats_disable;
 
 	if (WARN_ON_ONCE(status->flag & RX_FLAG_RADIOTAP_TLV_AT_END &&
 			 !skb_mac_header_was_set(origskb))) {
@@ -983,6 +1028,10 @@ ieee80211_rx_monitor(struct ieee80211_local *local, struct sk_buff *origskb,
 			if (skb) {
 				skb->dev = sdata->dev;
 				ieee80211_rx_stats(skb->dev, skb->len);
+				if (!tid_stats_disable) {
+					ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_TOTAL_PKTS);
+					ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_NETIF_PKTS);
+				}
 				netif_receive_skb(skb);
 			}
 		}
@@ -1565,7 +1614,7 @@ ieee80211_rx_h_check_dup(struct ieee80211_rx_data *rx)
 		     rx->sta->last_seq_ctrl[rx->seqno_idx] == hdr->seq_ctrl)) {
 		I802_DEBUG_INC(rx->local->dot11FrameDuplicateCount);
 		rx->link_sta->rx_stats.num_duplicates++;
-		return RX_DROP_U_DUP;
+		return RX_DROP_DUP;
 	} else if (!(status->flag & RX_FLAG_AMSDU_MORE)) {
 		rx->sta->last_seq_ctrl[rx->seqno_idx] = hdr->seq_ctrl;
 	}
@@ -1619,7 +1668,7 @@ ieee80211_rx_h_check(struct ieee80211_rx_data *rx)
 		    cfg80211_rx_spurious_frame(rx->sdata->dev,
 					       hdr->addr2,
 					       GFP_ATOMIC))
-			return RX_DROP_U_SPURIOUS;
+			return RX_DROP_SPURIOUS;
 
 		return RX_DROP_MONITOR;
 	}
@@ -1967,6 +2016,7 @@ ieee80211_rx_h_sta_process(struct ieee80211_rx_data *rx)
 					rx->sdata->dev, sta->sta.addr,
 					GFP_ATOMIC,
 					rx->link_id);
+			ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_UNEXPECTED_4ADDR_FRAME);
 			return RX_DROP_M_UNEXPECTED_4ADDR_FRAME;
 		}
 		/*
@@ -2071,8 +2121,10 @@ ieee80211_rx_h_decrypt(struct ieee80211_rx_data *rx)
 		    !(status->flag & RX_FLAG_IV_STRIPPED)) {
 			keyid = ieee80211_get_keyid(rx->skb);
 
-			if (unlikely(keyid < 0))
+			if (unlikely(keyid < 0)) {
+				ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_NO_KEY_ID);
 				return RX_DROP_U_NO_KEY_ID;
+			}
 
 			ptk_idx = rcu_dereference(rx->sta->ptk[keyid]);
 		}
@@ -2102,6 +2154,7 @@ ieee80211_rx_h_decrypt(struct ieee80211_rx_data *rx)
 				cfg80211_rx_unprot_mlme_mgmt(rx->sdata->dev,
 							     skb->data,
 							     skb->len);
+			ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_BCN_KEYIDX);
 			return RX_DROP_M_BAD_BCN_KEYIDX;
 		}
 
@@ -2115,8 +2168,10 @@ ieee80211_rx_h_decrypt(struct ieee80211_rx_data *rx)
 			return RX_CONTINUE;
 
 		if (mmie_keyidx < NUM_DEFAULT_KEYS ||
-		    mmie_keyidx >= NUM_DEFAULT_KEYS + NUM_DEFAULT_MGMT_KEYS)
+		    mmie_keyidx >= NUM_DEFAULT_KEYS + NUM_DEFAULT_MGMT_KEYS) {
+			ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_MGMT_KEYIDX);
 			return RX_DROP_M_BAD_MGMT_KEYIDX; /* unexpected BIP keyidx */
+		}
 		if (rx->link_sta) {
 			if (ieee80211_is_group_privacy_action(skb) &&
 			    test_sta_flag(rx->sta, WLAN_STA_MFP))
@@ -2176,8 +2231,10 @@ ieee80211_rx_h_decrypt(struct ieee80211_rx_data *rx)
 
 		keyidx = ieee80211_get_keyid(rx->skb);
 
-		if (unlikely(keyidx < 0))
+		if (unlikely(keyidx < 0)) {
+			ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_NO_KEY_ID);
 			return RX_DROP_U_NO_KEY_ID;
+		}
 
 		/* check per-station GTK first, if multicast packet */
 		if (is_multicast_ether_addr(hdr->addr1) && rx->link_sta)
@@ -2244,6 +2301,7 @@ ieee80211_rx_h_decrypt(struct ieee80211_rx_data *rx)
 		break;
 	default:
 		result = RX_DROP_U_BAD_CIPHER;
+		ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_CIPHER);
 	}
 
 	/* the hdr variable is invalid after the decrypt handlers */
@@ -2387,8 +2445,10 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 
 	I802_DEBUG_INC(rx->local->rx_handlers_fragments);
 
-	if (skb_linearize(rx->skb))
+	if (skb_linearize(rx->skb)) {
+		ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_OOM);
 		return RX_DROP_U_OOM;
+	}
 
 	/*
 	 *  skb_linearize() might change the skb->data and
@@ -2450,12 +2510,16 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 		int i;
 		u8 pn[IEEE80211_CCMP_PN_LEN], *rpn;
 
-		if (!requires_sequential_pn(rx, fc))
+		if (!requires_sequential_pn(rx, fc)) {
+			ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_NONSEQ_PN);
 			return RX_DROP_U_NONSEQ_PN;
+		}
 
 		/* Prevent mixed key and fragment cache attacks */
-		if (entry->key_color != rx->key->color)
+		if (entry->key_color != rx->key->color) {
+			ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_BAD_KEY_COLOR);
 			return RX_DROP_U_BAD_KEY_COLOR;
+		}
 
 		memcpy(pn, entry->last_pn, IEEE80211_CCMP_PN_LEN);
 		for (i = IEEE80211_CCMP_PN_LEN - 1; i >= 0; i--) {
@@ -2465,8 +2529,10 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 		}
 
 		rpn = rx->ccm_gcm.pn;
-		if (memcmp(pn, rpn, IEEE80211_CCMP_PN_LEN))
+		if (memcmp(pn, rpn, IEEE80211_CCMP_PN_LEN)) {
+			ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_REPLAY);
 			return RX_DROP_U_REPLAY;
+		}
 		memcpy(entry->last_pn, pn, IEEE80211_CCMP_PN_LEN);
 	} else if (entry->is_protected &&
 		   (!rx->key ||
@@ -2477,10 +2543,12 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 		 * if for TKIP Michael MIC should protect us, and WEP is a
 		 * lost cause anyway.
 		 */
+		ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_EXPECT_DEFRAG_PROT);
 		return RX_DROP_U_EXPECT_DEFRAG_PROT;
 	} else if (entry->is_protected && rx->key &&
 		   entry->key_color != rx->key->color &&
 		   (status->flag & RX_FLAG_DECRYPTED)) {
+		ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_BAD_KEY_COLOR);
 		return RX_DROP_U_BAD_KEY_COLOR;
 	}
 
@@ -2500,6 +2568,7 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 					      GFP_ATOMIC))) {
 			I802_DEBUG_INC(rx->local->rx_handlers_drop_defrag);
 			__skb_queue_purge(&entry->skb_list);
+			ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_OOM);
 			return RX_DROP_U_OOM;
 		}
 	}
@@ -2846,15 +2915,20 @@ static void
 ieee80211_deliver_skb(struct ieee80211_rx_data *rx)
 {
 	struct ieee80211_sub_if_data *sdata = rx->sdata;
+	struct ieee80211_local *local = rx->local;
 	struct net_device *dev = sdata->dev;
 	struct sk_buff *skb, *xmit_skb;
 	struct ethhdr *ehdr = (struct ethhdr *) rx->skb->data;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(rx->skb);
 	struct sta_info *dsta;
+	bool tid_stats_disable = local->hw.tid_stats_disable;
 
 	skb = rx->skb;
 	xmit_skb = NULL;
 
 	ieee80211_rx_stats(dev, skb->len);
+	if (!tid_stats_disable)
+		ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_TOTAL_PKTS);
 
 	if (rx->sta) {
 		/* The seqno index has the same property as needed
@@ -2878,10 +2952,13 @@ ieee80211_deliver_skb(struct ieee80211_rx_data *rx)
 			 * send multicast frames both to higher layers in
 			 * local net stack and back to the wireless medium
 			 */
+
 			xmit_skb = skb_copy(skb, GFP_ATOMIC);
 			if (!xmit_skb)
 				net_info_ratelimited("%s: failed to clone multicast frame\n",
 						    dev->name);
+			if (!tid_stats_disable)
+				ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_MULTICAST_PKTS);
 		} else if (!is_multicast_ether_addr(ehdr->h_dest) &&
 			   !ether_addr_equal(ehdr->h_dest, ehdr->h_source)) {
 			dsta = sta_info_get(sdata, ehdr->h_dest);
@@ -2895,6 +2972,8 @@ ieee80211_deliver_skb(struct ieee80211_rx_data *rx)
 				xmit_skb = skb;
 				skb = NULL;
 				atomic_inc(&rx->sta->rx_forwarded_pkts);
+				if (!tid_stats_disable)
+					ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_FORWARDED_PKTS);
 			}
 		}
 	}
@@ -2927,6 +3006,8 @@ ieee80211_deliver_skb(struct ieee80211_rx_data *rx)
 
 	if (skb) {
 		skb->protocol = eth_type_trans(skb, dev);
+		if (!tid_stats_disable)
+			ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_LOCAL_STACK_PKTS);
 		ieee80211_deliver_skb_to_local_stack(skb, rx);
 	}
 
@@ -3024,6 +3105,7 @@ ieee80211_rx_mesh_data(struct ieee80211_sub_if_data *sdata, struct sta_info *sta
 	struct ieee80211_hdr *fwd_hdr;
 	struct ieee80211s_hdr *mesh_hdr;
 	struct ieee80211_tx_info *info;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(skb);
 	struct sk_buff *fwd_skb;
 	struct ethhdr *eth;
 	bool multicast;
@@ -3133,11 +3215,15 @@ ieee80211_rx_mesh_data(struct ieee80211_sub_if_data *sdata, struct sta_info *sta
 		fwd_skb = skb;
 		skb = NULL;
 
-		if (skb_cow_head(fwd_skb, hdrlen - sizeof(struct ethhdr)))
+		if (skb_cow_head(fwd_skb, hdrlen - sizeof(struct ethhdr))) {
+			ieee80211_rx_drop_stats_reason(sdata, skb->len, status->tid, RX_DROP_OOM);
 			return RX_DROP_U_OOM;
+		}
 
-		if (skb_linearize(fwd_skb))
+		if (skb_linearize(fwd_skb)) {
+			ieee80211_rx_drop_stats_reason(sdata, skb->len, status->tid, RX_DROP_OOM);
 			return RX_DROP_U_OOM;
+		}
 	}
 
 	fwd_hdr = skb_push(fwd_skb, hdrlen - sizeof(struct ethhdr));
@@ -3198,6 +3284,7 @@ __ieee80211_rx_h_amsdu(struct ieee80211_rx_data *rx, u8 data_offset)
 {
 	struct net_device *dev = rx->sdata->dev;
 	struct sk_buff *skb = rx->skb;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(skb);
 	struct ieee80211_hdr *hdr = (struct ieee80211_hdr *)skb->data;
 	__le16 fc = hdr->frame_control;
 	struct sk_buff_head frame_list;
@@ -3232,8 +3319,10 @@ __ieee80211_rx_h_amsdu(struct ieee80211_rx_data *rx, u8 data_offset)
 	if (ieee80211_data_to_8023_exthdr(skb, &ethhdr,
 					  rx->sdata->vif.addr,
 					  rx->sdata->vif.type,
-					  data_offset, true))
+					  data_offset, true)) {
+		ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_AMSDU);
 		return RX_DROP_U_BAD_AMSDU;
+	}
 
 	if (rx->sta && rx->sta->amsdu_mesh_control < 0) {
 		s8 valid = -1;
@@ -3274,8 +3363,10 @@ __ieee80211_rx_h_amsdu(struct ieee80211_rx_data *rx, u8 data_offset)
 			goto free;
 		}
 
-		if (!ieee80211_frame_allowed(rx, fc))
+		if (!ieee80211_frame_allowed(rx, fc)) {
+			ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_FRAME_ERR);
 			goto free;
+		}
 
 		ieee80211_deliver_skb(rx);
 		continue;
@@ -3317,12 +3408,15 @@ ieee80211_rx_h_amsdu(struct ieee80211_rx_data *rx)
 		case NL80211_IFTYPE_MESH_POINT:
 			break;
 		default:
+			ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_4ADDR);
 			return RX_DROP_U_BAD_4ADDR;
 		}
 	}
 
-	if (is_multicast_ether_addr(hdr->addr1) || !rx->sta)
+	if (is_multicast_ether_addr(hdr->addr1) || !rx->sta) {
+		ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_AMSDU);
 		return RX_DROP_U_BAD_AMSDU;
+	}
 
 	if (rx->key) {
 		/*
@@ -3335,6 +3429,7 @@ ieee80211_rx_h_amsdu(struct ieee80211_rx_data *rx)
 		case WLAN_CIPHER_SUITE_WEP40:
 		case WLAN_CIPHER_SUITE_WEP104:
 		case WLAN_CIPHER_SUITE_TKIP:
+			ieee80211_rx_drop_stats_reason(rx->sdata, skb->len, status->tid, RX_DROP_BAD_AMSDU_CIPHER);
 			return RX_DROP_U_BAD_AMSDU_CIPHER;
 		default:
 			break;
@@ -3351,6 +3446,7 @@ ieee80211_rx_h_data(struct ieee80211_rx_data *rx)
 	struct ieee80211_local *local = rx->local;
 	struct net_device *dev = sdata->dev;
 	struct ieee80211_hdr *hdr = (struct ieee80211_hdr *)rx->skb->data;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(rx->skb);
 	__le16 fc = hdr->frame_control;
 	ieee80211_rx_result res;
 	bool port_control;
@@ -3363,8 +3459,10 @@ ieee80211_rx_h_data(struct ieee80211_rx_data *rx)
 		return RX_DROP_MONITOR;
 
 	err = __ieee80211_data_to_8023(rx, &port_control);
-	if (unlikely(err))
+	if (unlikely(err)) {
+		ieee80211_rx_drop_stats_reason(sdata, rx->skb->len, status->tid, RX_DROP_INVALID_8023);
 		return RX_DROP_U_INVALID_8023;
+	}
 
 	/*
 	 * Send unexpected-4addr-frame event to hostapd. For older versions,
@@ -3407,6 +3505,7 @@ ieee80211_rx_h_data(struct ieee80211_rx_data *rx)
 		    (tf->action_code == WLAN_TDLS_CHANNEL_SWITCH_REQUEST ||
 		     tf->action_code == WLAN_TDLS_CHANNEL_SWITCH_RESPONSE)) {
 			rx->skb->protocol = cpu_to_be16(ETH_P_TDLS);
+			ieee80211_rx_stats_reason(sdata, rx->skb->len, status->tid, RX_QUEUED_PKTS);
 			__ieee80211_queue_skb_to_iface(sdata, rx->link_id,
 						       rx->sta, rx->skb);
 			return RX_QUEUED;
@@ -3618,8 +3717,10 @@ ieee80211_rx_h_mgmt_check(struct ieee80211_rx_data *rx)
 
 	/* drop too small action frames */
 	if (ieee80211_is_action(mgmt->frame_control) &&
-	    rx->skb->len < IEEE80211_MIN_ACTION_SIZE)
+	    rx->skb->len < IEEE80211_MIN_ACTION_SIZE) {
+		ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_RUNT_ACTION);
 		return RX_DROP_U_RUNT_ACTION;
+	}
 
 	if (rx->sdata->vif.type == NL80211_IFTYPE_AP &&
 	    ieee80211_is_beacon(mgmt->frame_control) &&
@@ -3640,8 +3741,11 @@ ieee80211_rx_h_mgmt_check(struct ieee80211_rx_data *rx)
 		rx->flags |= IEEE80211_RX_BEACON_REPORTED;
 	}
 
-	if (ieee80211_drop_unencrypted_mgmt(rx))
+	if (ieee80211_drop_unencrypted_mgmt(rx)) {
+		ieee80211_rx_drop_stats_reason(rx->sdata, rx->skb->len, status->tid, RX_DROP_UNPROT_ACTION);
+
 		return RX_DROP_U_UNPROT_ACTION;
+	}
 
 	return RX_CONTINUE;
 }
@@ -3713,8 +3817,10 @@ ieee80211_rx_h_action(struct ieee80211_rx_data *rx)
 
 	if (!rx->sta && mgmt->u.action.category != WLAN_CATEGORY_PUBLIC &&
 	    mgmt->u.action.category != WLAN_CATEGORY_SELF_PROTECTED &&
-	    mgmt->u.action.category != WLAN_CATEGORY_SPECTRUM_MGMT)
+	    mgmt->u.action.category != WLAN_CATEGORY_SPECTRUM_MGMT) {
+		ieee80211_rx_drop_stats_reason(sdata, len, status->tid, RX_DROP_ACTION_UNKNOWN_SRC);
 		return RX_DROP_U_ACTION_UNKNOWN_SRC;
+	}
 
 	switch (mgmt->u.action.category) {
 	case WLAN_CATEGORY_HT:
@@ -4010,6 +4116,7 @@ ieee80211_rx_h_action(struct ieee80211_rx_data *rx)
 
  queue:
 	ieee80211_queue_skb_to_iface(sdata, rx->link_id, rx->sta, rx->skb);
+	ieee80211_rx_stats_reason(sdata, rx->skb->len, status->tid, RX_QUEUED_PKTS);
 	return RX_QUEUED;
 }
 
@@ -4138,8 +4245,11 @@ ieee80211_rx_h_action_return(struct ieee80211_rx_data *rx)
 		return RX_DROP_MONITOR;
 
 	/* do not return rejected action frames */
-	if (mgmt->u.action.category & 0x80)
+	if (mgmt->u.action.category & 0x80) {
+		ieee80211_rx_drop_stats_reason(sdata, rx->skb->len, status->tid,
+					       RX_DROP_REJECTED_ACTION_RESPONSE);
 		return RX_DROP_U_REJECTED_ACTION_RESPONSE;
+	}
 
 	nskb = skb_copy_expand(rx->skb, local->hw.extra_tx_headroom, 0,
 			       GFP_ATOMIC);
@@ -4175,6 +4285,7 @@ ieee80211_rx_h_ext(struct ieee80211_rx_data *rx)
 {
 	struct ieee80211_sub_if_data *sdata = rx->sdata;
 	struct ieee80211_hdr *hdr = (void *)rx->skb->data;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(rx->skb);
 
 	if (!ieee80211_is_ext(hdr->frame_control))
 		return RX_CONTINUE;
@@ -4182,6 +4293,7 @@ ieee80211_rx_h_ext(struct ieee80211_rx_data *rx)
 	if (sdata->vif.type != NL80211_IFTYPE_STATION)
 		return RX_DROP_MONITOR;
 
+	ieee80211_rx_stats_reason(sdata, rx->skb->len, status->tid, RX_QUEUED_PKTS);
 	/* for now only beacons are ext, so queue them */
 	ieee80211_queue_skb_to_iface(sdata, rx->link_id, rx->sta, rx->skb);
 
@@ -4193,6 +4305,7 @@ ieee80211_rx_h_mgmt(struct ieee80211_rx_data *rx)
 {
 	struct ieee80211_sub_if_data *sdata = rx->sdata;
 	struct ieee80211_mgmt *mgmt = (void *)rx->skb->data;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(rx->skb);
 	__le16 stype;
 
 	stype = mgmt->frame_control & cpu_to_le16(IEEE80211_FCTL_STYPE);
@@ -4240,6 +4353,7 @@ ieee80211_rx_h_mgmt(struct ieee80211_rx_data *rx)
 		return RX_DROP_MONITOR;
 	}
 
+	ieee80211_rx_stats_reason(sdata, rx->skb->len, status->tid, RX_QUEUED_PKTS);
 	ieee80211_queue_skb_to_iface(sdata, rx->link_id, rx->sta, rx->skb);
 
 	return RX_QUEUED;
@@ -4255,6 +4369,7 @@ static void ieee80211_rx_cooked_monitor(struct ieee80211_rx_data *rx,
 	struct net_device *prev_dev = NULL;
 	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(skb);
 	int needed_headroom;
+	bool tid_stats_disable = local->hw.tid_stats_disable;
 
 	/*
 	 * If cooked monitor has been processed already, then
@@ -4302,6 +4417,10 @@ static void ieee80211_rx_cooked_monitor(struct ieee80211_rx_data *rx,
 
 		prev_dev = sdata->dev;
 		ieee80211_rx_stats(sdata->dev, skb->len);
+		if (!tid_stats_disable) {
+			ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_TOTAL_PKTS);
+			ieee80211_rx_stats_reason(sdata, skb->len, status->tid, RX_MONITOR_PKTS);
+		}
 	}
 
 	if (prev_dev) {
@@ -5022,11 +5141,13 @@ static void ieee80211_rx_8023(struct ieee80211_rx_data *rx,
 {
 	struct ieee80211_sta_rx_stats *stats;
 	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(rx->skb);
+	struct ieee80211_local *local = rx->local;
 	struct sta_info *sta = rx->sta;
 	struct link_sta_info *link_sta;
 	struct sk_buff *skb = rx->skb;
 	void *sa = skb->data + ETH_ALEN;
 	void *da = skb->data;
+	bool tid_stats_disable = local->hw.tid_stats_disable;
 
 	if (rx->link_id >= 0) {
 		link_sta = rcu_dereference(sta->link[rx->link_id]);
@@ -5077,6 +5198,8 @@ static void ieee80211_rx_8023(struct ieee80211_rx_data *rx,
 	skb->dev = fast_rx->dev;
 
 	ieee80211_rx_stats(fast_rx->dev, skb->len);
+	if (!tid_stats_disable)
+		ieee80211_rx_stats_reason(rx->sdata, skb->len, status->tid, RX_TOTAL_PKTS);
 
 	/* The seqno index has the same property as needed
 	 * for the rx_msdu field, i.e. it is IEEE80211_NUM_TIDS
@@ -5092,6 +5215,8 @@ static void ieee80211_rx_8023(struct ieee80211_rx_data *rx,
 		struct sk_buff *xmit_skb = NULL;
 		if (is_multicast_ether_addr(da)) {
 			xmit_skb = skb_copy(skb, GFP_ATOMIC);
+			if (!tid_stats_disable)
+				ieee80211_rx_stats_reason(rx->sdata, skb->len, status->tid, RX_MULTICAST_PKTS);
 		} else if (!ether_addr_equal(da, sa) &&
 			   sta_info_get(rx->sdata, da)) {
 			xmit_skb = skb;
@@ -5110,6 +5235,8 @@ static void ieee80211_rx_8023(struct ieee80211_rx_data *rx,
 			skb_reset_mac_header(xmit_skb);
 			dev_queue_xmit(xmit_skb);
 			atomic_inc(&rx->sta->rx_forwarded_pkts);
+			if (!tid_stats_disable)
+				ieee80211_rx_stats_reason(rx->sdata, skb->len, status->tid, RX_FORWARDED_PKTS);
 		}
 
 		if (!skb)
@@ -5119,11 +5246,15 @@ static void ieee80211_rx_8023(struct ieee80211_rx_data *rx,
 #ifdef CPTCFG_MAC80211_PPE_SUPPORT
 	if (ieee80211_netif_rx_ppe(rx, skb)) {
 		atomic_inc(&sta->rx_netif_pkts);
+		if (!tid_stats_disable)
+			ieee80211_rx_stats_reason(rx->sdata, skb->len, status->tid, RX_NETIF_PKTS);
 		return;
 	}
 #endif
 	/* deliver to local stack */
 	skb->protocol = eth_type_trans(skb, fast_rx->dev);
+	if (!tid_stats_disable)
+		ieee80211_rx_stats_reason(rx->sdata, skb->len, status->tid, RX_LOCAL_STACK_PKTS);
 	ieee80211_deliver_skb_to_local_stack(skb, rx);
 }
 
@@ -5288,6 +5419,8 @@ static bool ieee80211_prepare_and_rx_handle(struct ieee80211_rx_data *rx,
 	struct ieee80211_hdr *hdr = (void *)skb->data;
 	struct link_sta_info *link_sta = rx->link_sta;
 	struct ieee80211_link_data *link = rx->link;
+	struct ieee80211_rx_status *status = IEEE80211_SKB_RXCB(skb);
+	bool tid_stats_disable = local->hw.tid_stats_disable;
 
 	rx->skb = skb;
 
@@ -5302,12 +5435,17 @@ static bool ieee80211_prepare_and_rx_handle(struct ieee80211_rx_data *rx,
 		struct ieee80211_fast_rx *fast_rx;
 
 		fast_rx = rcu_dereference(rx->sta->fast_rx);
-		if (fast_rx && ieee80211_invoke_fast_rx(rx, fast_rx))
+		if (fast_rx && ieee80211_invoke_fast_rx(rx, fast_rx)) {
+			if (!tid_stats_disable)
+				ieee80211_rx_stats_reason(sdata, skb->len, status->tid,  RX_FAST_PATH_PKTS);
 			return true;
+		}
 	}
 
-	if (!ieee80211_accept_frame(rx))
+	if (!ieee80211_accept_frame(rx)) {
+		ieee80211_rx_drop_stats_reason(sdata, skb->len, status->tid, RX_DROP_INVALID_SKB);
 		return false;
+	}
 
 	if (!consume) {
 		struct skb_shared_hwtstamps *shwt;
@@ -5373,6 +5511,8 @@ static void __ieee80211_rx_handle_8023(struct ieee80211_hw *hw,
 	rx.list = list;
 	rx.link_id = -1;
 
+	status->tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
+
 	I802_DEBUG_INC(local->dot11ReceivedFragmentCount);
 
 	/* drop frame if too short for header */
@@ -5393,12 +5533,16 @@ static void __ieee80211_rx_handle_8023(struct ieee80211_hw *hw,
 	 * the deflink is fine?
 	 */
 	sta = container_of(pubsta, struct sta_info, sta);
-	if (!ieee80211_rx_data_set_sta(&rx, sta, link_id, false))
+	if (!ieee80211_rx_data_set_sta(&rx, sta, link_id, false)) {
+		ieee80211_rx_drop_stats_reason(sta->sdata, skb->len, status->tid, RX_DROP_NO_SET_STA);
 		goto drop;
+	}
 
 	fast_rx = rcu_dereference(rx.sta->fast_rx);
-	if (!fast_rx)
+	if (!fast_rx) {
+		ieee80211_rx_drop_stats_reason(sta->sdata, skb->len, status->tid, RX_DROP_NO_FAST_RX);
 		goto drop;
+	}
 
 	ieee80211_rx_8023(&rx, fast_rx, skb->len);
 	return;
@@ -5476,6 +5620,7 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 	bool prev_flag;
 	bool is_mgmt = false;
 
+	hdr = (struct ieee80211_hdr *)skb->data;
 	fc = ((struct ieee80211_hdr *)skb->data)->frame_control;
 	memset(&rx, 0, sizeof(rx));
 	rx.skb = skb;
@@ -5483,6 +5628,11 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 	rx.list = list;
 	rx.link_id = -1;
 
+	if (ieee80211_is_data_qos(fc))
+		status->tid = ieee80211_get_tid(hdr);
+	else
+		status->tid = skb->priority & IEEE80211_QOS_CTL_TAG1D_MASK;
+
 	if (ieee80211_is_data(fc) || ieee80211_is_mgmt(fc))
 		I802_DEBUG_INC(local->dot11ReceivedFragmentCount);
 
@@ -5538,8 +5688,10 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 
 				link_sta = link_sta_info_get_bss(rx.sdata,
 								 hdr->addr2);
-				if (!link_sta)
+				if (!link_sta) {
+					ieee80211_rx_drop_stats_reason(rx.sdata, skb->len, status->tid, RX_DROP_NO_LINK_STA);
 					goto out;
+				}
 
 				ieee80211_rx_data_set_link(&rx, link_sta->link_id);
 			}
@@ -5558,8 +5710,10 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 			}
 
 			rx.sdata = prev_sta->sdata;
-			if (!ieee80211_rx_data_set_sta(&rx, prev_sta, link_id, only_monitor))
+			if (!ieee80211_rx_data_set_sta(&rx, prev_sta, link_id, only_monitor)) {
+				ieee80211_rx_drop_stats_reason(rx.sdata, skb->len, status->tid, RX_DROP_NO_SET_STA);
 				goto out;
+			}
 
 			if (!status->link_valid && prev_sta->sta.mlo)
 				continue;
@@ -5571,11 +5725,15 @@ static void __ieee80211_rx_handle_packet(struct ieee80211_hw *hw,
 
 		if (prev_sta) {
 			rx.sdata = prev_sta->sdata;
-			if (!ieee80211_rx_data_set_sta(&rx, prev_sta, link_id, only_monitor))
+			if (!ieee80211_rx_data_set_sta(&rx, prev_sta, link_id, only_monitor)) {
+				ieee80211_rx_drop_stats_reason(rx.sdata, skb->len, status->tid, RX_DROP_NO_SET_STA);
 				goto out;
+			}
 
-			if (!status->link_valid && prev_sta->sta.mlo)
+			if (!status->link_valid && prev_sta->sta.mlo) {
+				ieee80211_rx_drop_stats_reason(rx.sdata, skb->len, status->tid, RX_DROP_NO_LINK_STA);
 				goto out;
+			}
 
 			if (ieee80211_prepare_and_rx_handle(&rx, skb, true))
 				return;
diff --git a/net/mac80211/wpa.c b/net/mac80211/wpa.c
index 6978de8..21ae441 100644
--- a/net/mac80211/wpa.c
+++ b/net/mac80211/wpa.c
@@ -90,7 +90,6 @@ ieee80211_tx_h_michael_mic_add(struct ieee80211_tx_data *tx)
 	return TX_CONTINUE;
 }
 
-
 ieee80211_rx_result
 ieee80211_rx_h_michael_mic_verify(struct ieee80211_rx_data *rx)
 {
-- 
2.34.1

